pallets/farming/src/gauge.rs

```diff
@@ -0,0 +1,494 @@
+            gauge_amount: Default::default(),
+            total_time_factor: Default::default(),
+            latest_time_factor: Default::default(),
+            claimed_time_factor: Default::default(),
+            gauge_start_block: Default::default(),
+            gauge_stop_block: Default::default(),
+            gauge_last_block: Default::default(),
+            last_claim_block: Default::default(),
+        }
+    }
+}
+
+/// The Reward Pool Info.
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct GaugePoolInfo<BalanceOf: HasCompact, CurrencyIdOf: Ord, AccountIdOf, BlockNumberFor> {
+    pub pid: PoolId,
+    pub token: CurrencyIdOf,
+    pub keeper: AccountIdOf,
+    pub reward_issuer: AccountIdOf,
+    pub rewards: BTreeMap<CurrencyIdOf, (BalanceOf, BalanceOf, BalanceOf)>,
+    pub gauge_basic_rewards: BTreeMap<CurrencyIdOf, BalanceOf>,
+    pub max_block: BlockNumberFor,
+    pub gauge_amount: BalanceOf,
+    pub total_time_factor: u128,
+    pub gauge_state: GaugeState,
+    pub gauge_last_block: BlockNumberFor,
+}
+
+#[derive(Encode, Decode, Copy, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub enum GaugeState {
+    Unbond,
+    Bonded,
+}
+
+impl<BalanceOf, CurrencyIdOf, AccountIdOf, BlockNumberFor>
+    GaugePoolInfo<BalanceOf, CurrencyIdOf, AccountIdOf, BlockNumberFor>
+where
+    BalanceOf: Default + HasCompact,
+    CurrencyIdOf: Ord + Default,
+    BlockNumberFor: Clone,
+{
+    pub fn new(
+        pid: PoolId,
+        token: CurrencyIdOf,
+        keeper: AccountIdOf,
+        reward_issuer: AccountIdOf,
+        gauge_basic_rewards: BTreeMap<CurrencyIdOf, BalanceOf>,
+        max_block: BlockNumberFor,
+        current_block_number: BlockNumberFor,
+    ) -> Self {
+        Self {
+            pid,
+            token,
+            keeper,
+            reward_issuer,
+            rewards: BTreeMap::new(),
+            gauge_basic_rewards,
+            max_block,
+            gauge_amount: Default::default(),
+            total_time_factor: Default::default(),
+            gauge_last_block: current_block_number,
+            gauge_state: GaugeState::Bonded,
+        }
+    }
+}
+
+impl<T: Config> Pallet<T>
+where
+    BlockNumberFor<T>: AtLeast32BitUnsigned + Copy,
+    BalanceOf<T>: AtLeast32BitUnsigned + Copy,
+{
+    pub fn create_gauge_pool(
+        pid: PoolId,
+        pool_info: &mut PoolInfo<BalanceOf<T>, CurrencyIdOf<T>, AccountIdOf<T>, BlockNumberFor<T>>,
+        gauge_token: CurrencyIdOf<T>,
+        gauge_basic_rewards: BTreeMap<CurrencyIdOf<T>, BalanceOf<T>>,
+        max_block: BlockNumberFor<T>,
+    ) -> DispatchResult {
+        let gid = Self::gauge_pool_next_id();
+        pool_info.gauge = Some(gid);
+        let current_block_number = frame_system::Pallet::<T>::block_number();
+        let gauge_pool_info = GaugePoolInfo::new(
+            pid,
+            gauge_token,
+            pool_info.keeper.clone(),
+            pool_info.reward_issuer.clone(),
```

Comment:
dont think these need to be cloned

Reply:
There's a mutation operation before this clone, need clone, otherwise has `Cannot move` error.

Reply:
Yeah my bad.
Just wondering though, can we implement Copy on the Accoutnid type, it's just a [u8;32] after all